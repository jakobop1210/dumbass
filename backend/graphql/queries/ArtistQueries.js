const { ArtistType } = require('../types/ArtistType')
const {
    GraphQLID,
    GraphQLList,
    GraphQLNonNull,
    GraphQLInt,
    GraphQLString
} = require('graphql')
const Artist = require('../../models/Artist')

/**
 * GraphQL Query Field - getArtistById.
 *
 * Retrieves a single artist by their unique ID. Returns an `ArtistType` object corresponding to the provided ID.
 *
 * @type {Object} GraphQL query configuration object.
 * @property {ArtistType} type - The GraphQL type that this query will return.
 * @property {Object} args - Arguments required for this query.
 * @property {function} resolve - The resolver function to execute the query.
 */
const getArtistById = {
    type: ArtistType,
    args: {
        id: { type: GraphQLID }
    },
    resolve: async (parent, args) => {
        try {
            const artist = await Artist.findOne({ id: Number(args.id) })
            if (!artist) {
                throw new Error(`Artist with id ${args.id} not found.`)
            }
            return artist
        } catch (error) {
            throw new Error(error.message)
        }
    }
}

/**
 * GraphQL Query Field - getArtistsByIds.
 *
 * Retrieves multiple artists based on an array of their IDs. Returns a list of `ArtistType` objects.
 *
 * @type {Object} GraphQL query configuration object.
 * @property {GraphQLList(ArtistType)} type - The GraphQL type that this query will return.
 * @property {Object} args - Arguments required for this query.
 * @property {function} resolve - The resolver function to execute the query.
 */
const getArtistsByIds = {
    type: new GraphQLList(ArtistType),
    args: {
        ids: { type: new GraphQLNonNull(new GraphQLList(GraphQLInt)) }
    },
    resolve: async (parent, args) => {
        try {
            const artists = await Artist.find({
                id: { $in: args.ids }
            })
            if (!artists) {
                throw new Error(`Artists with ids ${args.ids} not found.`)
            }
            return artists
        } catch (error) {
            throw new Error(error.message)
        }
    }
}

/**
 * GraphQL Query Field - countArtists.
 *
 * Counts the number of artists that match a given criteria, such as a name pattern. Returns an integer count.
 *
 * @type {Object} GraphQL query configuration object.
 * @property {GraphQLInt} type - The GraphQL type that this query will return (integer in this case).
 * @property {Object} args - Arguments required for this query.
 * @property {function} resolve - The resolver function to execute the query.
 */
const countArtists = {
    type: GraphQLInt,
    args: {
        name: { type: GraphQLString }
    },
    resolve: async (parent, args) => {
        let query = {}
        if (args.name) {
            query.name = new RegExp(args.name, 'i')
        }
        try {
            return await Artist.countDocuments(query)
        } catch (error) {
            throw new Error('Error counting artists. ' + error)
        }
    }
}

module.exports = { getArtistById, getArtistsByIds, countArtists }
