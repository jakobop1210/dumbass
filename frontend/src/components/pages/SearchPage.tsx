import { useEffect, useState } from 'react'
import { ArtistCardProps, SongCardProps } from '../molecules/ArtistSongCard'
import CardView from '../organisms/CardView'
import SearchBar from '../molecules/SearchBar'
import Breadcrumb from '../atoms/Breadcrumb'
import { useLocation } from 'react-router-dom'
import { useQuery } from '@apollo/client'
import {
    COUNT_ARTISTS,
    GET_ARTISTS_ON_NAME
} from '../../graphql/queries/artistQueries'
import {
    COUNT_SONGS,
    GET_SONGS_ON_TITLE
} from '../../graphql/queries/songQueries'
import Pagination from '../molecules/Pagination'
import { Artist, Song } from '../../lib/types'
import { useSearchParams } from 'react-router-dom'
import Dropdown from '../atoms/Dropdown'
import { ClipLoader } from 'react-spinners'

type CardProps = ArtistCardProps | SongCardProps

/**
 * SearchPage component to render and handle search functionality,
 * filtering and sorting for artists and songs.
 */
function SearchPage() {
    const location = useLocation()
    const queryParams = new URLSearchParams(location.search)

    const term = queryParams.get('term')
    const filter = queryParams.get('filter')
    const sort = queryParams.get('sort')
    const [searchParams, setSearchParams] = useSearchParams()

    const validFilters = ['song', 'artist']
    const defaultFilter = 'song'
    const validSort = ['relevance', 'rating', 'alphabetical']
    const defaultSort = 'relevance'

    const filterFromURL = searchParams.get('filter') || defaultFilter
    const sortFromURL = searchParams.get('sort') || defaultSort

    const { data: totalSongsData } = useQuery(COUNT_SONGS, {
        variables: { title: term },
        skip: filter !== 'song'
    })

    const { data: totalArtistsData } = useQuery(COUNT_ARTISTS, {
        variables: { name: term },
        skip: filter !== 'artist'
    })

    const [selectedFilter, setSelectedFilter] = useState(
        validFilters.includes(filterFromURL) ? filterFromURL : defaultFilter
    )
    const [selectedSort, setSelectedSorting] = useState(
        validSort.includes(sortFromURL) ? sortFromURL : defaultSort
    )

    useEffect(() => {
        const newSearchParams = new URLSearchParams(searchParams)
        if (searchParams.get('filter') !== selectedFilter) {
            newSearchParams.set('filter', selectedFilter)
        }
        if (searchParams.get('sort') !== selectedSort) {
            newSearchParams.set('sort', selectedSort)
        }
        setSearchParams(newSearchParams, { replace: true })
    }, [selectedFilter, selectedSort, searchParams, setSearchParams])

    const [data, setData] = useState<CardProps[]>([])
    const [currentPage, setCurrentPage] = useState(1)
    const [totalPages, setTotalPages] = useState(0)
    const itemsPerPage = 12

    const { data: artistsData, loading: artistsLoading } = useQuery(
        GET_ARTISTS_ON_NAME,
        {
            variables: {
                name: term,
                limit: itemsPerPage,
                sort: sort,
                page: currentPage
            },
            skip: filter !== 'artist',
            fetchPolicy: 'cache-and-network'
        }
    )

    const { data: songsData, loading: songsLoading } = useQuery(
        GET_SONGS_ON_TITLE,
        {
            variables: {
                title: term,
                limit: itemsPerPage,
                sort: sort,
                page: currentPage
            },
            skip: filter !== 'song',
            fetchPolicy: 'cache-and-network'
        }
    )

    useEffect(() => {
        if (filter === 'artist' && artistsData) {
            const artistCardData: ArtistCardProps[] =
                artistsData.getArtistsOnName.map((artist: Artist) => ({
                    cardType: 'artist',
                    id: artist.id,
                    title: artist.name,
                    alternateNames: artist.alternate_names,
                    imageUrl: artist.image_url,
                    rating: artist.average_rating,
                    numOfRatings: artist.number_of_ratings
                }))
            setData(artistCardData)
        } else if (filter === 'song' && songsData) {
            const songCards: SongCardProps[] = songsData.getSongsOnTitle.map(
                (song: Song) => ({
                    cardType: 'song',
                    id: song.id,
                    title: song.title,
                    artist: song.artist_names,
                    imageUrl: song.header_image_url,
                    rating: song.average_rating,
                    numOfRatings: song.number_of_ratings,
                    releaseDate: song.release_date
                })
            )
            setData(songCards)
        }
    }, [filter, artistsData, songsData])

    useEffect(() => {
        let totalItems = 0
        if (filter === 'artist' && totalArtistsData) {
            totalItems = totalArtistsData.countArtists
        } else if (filter === 'song' && totalSongsData) {
            totalItems = totalSongsData.countSongs
        }
        setTotalPages(Math.ceil(totalItems / itemsPerPage))
    }, [totalArtistsData, totalSongsData, itemsPerPage, filter])

    useEffect(() => {
        setCurrentPage(1)
    }, [term, selectedFilter, selectedSort])

    const isTermProvided = term !== null && term.trim() !== ''

    if (!isTermProvided) {
        return (
            <main className='w-full'>
                <Breadcrumb items={[{ name: 'Search' }]} />
                <header className='flex justify-center'>
                    <SearchBar
                        filterOptions={validFilters}
                        selectedFilter={selectedFilter}
                        selectedSort={selectedSort}
                        onFilterChange={(newFilter) =>
                            setSelectedFilter(newFilter)
                        }
                    />
                </header>
                <section className='flex text-xl justify-center w-full text-gray-500'>
                    Please enter a search term...
                </section>
            </main>
        )
    }

    return (
        <main className='w-full'>
            <Breadcrumb
                items={[
                    {
                        name: 'Search'
                    }
                ]}
            />
            <header className='flex justify-center'>
                <SearchBar
                    filterOptions={validFilters}
                    selectedFilter={selectedFilter}
                    selectedSort={selectedSort}
                    onFilterChange={(newFilter) => setSelectedFilter(newFilter)}
                />
            </header>
            <section className='flex justify-center mt-5 mb-10'>
                {data.length > 0 && (
                    <Dropdown
                        selectedFilter={selectedSort}
                        filterOptions={validSort}
                        onFilterChange={(newSorting) =>
                            setSelectedSorting(newSorting)
                        }
                        outsideSearchBar
                        title='Sort by'
                    />
                )}
            </section>
            <section className='w-full flex flex-col justify-center items-center'>
                <section className='w-full flex justify-center'>
                    {artistsLoading || songsLoading ? (
                        <div className='absolute h-96 flex items-center'>
                            <ClipLoader color={'#8fc0a9'} size={100} />
                        </div>
                    ) : (
                        <CardView cardData={data} />
                    )}
                </section>
                <Pagination
                    onClickPrevious={() => setCurrentPage(currentPage - 1)}
                    onClickNext={() => setCurrentPage(currentPage + 1)}
                    currentPage={currentPage}
                    totalPages={totalPages}
                    isLoading={artistsLoading || songsLoading}
                />
            </section>
        </main>
    )
}

export default SearchPage
