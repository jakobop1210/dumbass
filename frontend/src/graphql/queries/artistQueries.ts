import { gql } from '@apollo/client'

export const GET_TOP_ARTISTS = gql`
    query ($limit: Int) {
        getTopArtists(limit: $limit) {
            id
            name
            alternate_names
            description
            image_url
            average_rating
            number_of_ratings
        }
    }
`

export const GET_ARTIST_BY_ID = gql`
    query ($id: ID!) {
        getArtistById(id: $id) {
            alternate_names
            description
            image_url
            name
            average_rating
            number_of_ratings
        }
    }
`

export const GET_ARTISTS_BY_ID = gql`
    query ($ids: [ID!]!) {
        getArtistsByIds(ids: $ids) {
            id
            name
            alternate_names
            description
            image_url
            average_rating
            number_of_ratings
        }
    }
`
