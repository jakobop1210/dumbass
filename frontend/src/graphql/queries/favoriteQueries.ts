import { gql } from '@apollo/client'

/**
 * GraphQL query: CHECK_IF_FAVORITE.
 *
 * This query checks if a specific item (song or artist) is marked as a favorite by a user.
 * It requires the username of the user, the type of the item (song or artist), and the item's ID.
 *
 * @param {string} $username - The username of the user.
 * @param {string} $type - The type of the item ('song' or 'artist').
 * @param {number} $targetId - The ID of the item (song or artist).
 * @returns {boolean} A boolean indicating whether the item is a favorite of the user.
 */
export const CHECK_IF_FAVORITE = gql`
    query ($username: String!, $type: String!, $targetId: Int!) {
        checkIfFavorite(username: $username, type: $type, targetId: $targetId)
    }
`

/**
 * GraphQL query: GET_FAVORITES.
 *
 * This query retrieves a list of favorite items (songs or artists) for a specific user.
 * It requires the username of the user and returns an array of objects, each representing a favorite item.
 * Each object includes the type of the item (song or artist) and its ID.
 *
 * @param {string} $username - The username of the user whose favorites are being requested.
 * @returns {Array<Object>} An array of objects, each containing the type ('song' or 'artist') and the targetId of the favorite item.
 */
export const GET_FAVORITES = gql`
    query ($username: String!) {
        getFavorites(username: $username) {
            type
            targetId
        }
    }
`
