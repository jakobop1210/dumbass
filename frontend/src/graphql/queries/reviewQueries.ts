import { gql } from '@apollo/client'

/**
 * GraphQL query: GET_REVIEWS_BY_TARGET_ID.
 *
 * This query fetches reviews for a specific target item, identified by its type and ID.
 * The target can be either a song or an artist. The query returns a list of reviews, 
 * each containing the username of the reviewer, the content of the review, and the rating given.
 *
 * @param {string} $targetType - The type of the target for which reviews are being fetched ('song' or 'artist').
 * @param {number} $targetId - The unique identifier of the target (song or artist).
 * @returns {Array<Object>} An array of review objects, each containing the userName, content, and rating of the review.
 */
export const GET_REVIEWS_BY_TARGET_ID = gql`
    query ($targetType: String!, $targetId: Int!) {
        getReviewsByTarget(targetType: $targetType, targetId: $targetId) {
            userName
            content
            rating
        }
    }
`
