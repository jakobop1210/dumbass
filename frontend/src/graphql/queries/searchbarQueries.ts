import { gql } from '@apollo/client'

export const SEARCHBAR_DROPDOWN = gql`
    query ($searchType: String!, $searchString: String!, $limit: Int) {
        searchSearchbar(
            searchType: $searchType
            searchString: $searchString
            limit: $limit
        ) {
            ... on Artist {
                name
                id
            }
            ... on Song {
                title
                id
            }
        }
    }
`
