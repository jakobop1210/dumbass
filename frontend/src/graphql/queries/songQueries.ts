import { gql } from '@apollo/client'

/**
 * GraphQL query: GET_SONG_BY_ID.
 *
 * This query fetches detailed information about a song by its unique ID. 
 * The details include the song's lyrics, header image URL, release date, title, artist names, average rating, and the number of ratings.
 *
 * @param {string} $id - The unique ID of the song to fetch.
 * @returns {Object} An object containing detailed information about the song.
 */
export const GET_SONG_BY_ID = gql`
    query ($id: ID!) {
        getSongById(id: $id) {
            lyrics
            header_image_url
            release_date
            title
            artist_names
            average_rating
            number_of_ratings
        }
    }
`

/**
 * GraphQL query: GET_SONGS_BY_ID.
 *
 * This query retrieves detailed information for a list of songs based on their unique IDs.
 * The details include the song's ID, title, artist names, header image URL, release date, primary artist ID, average rating, and number of ratings.
 *
 * @param {Array<number>} $ids - An array of IDs of the songs to fetch.
 * @returns {Array<Object>} An array of objects, each containing detailed information about a song.
 */
export const GET_SONGS_BY_ID = gql`
    query ($ids: [Int]!) {
        getSongsByIds(ids: $ids) {
            id
            title
            artist_names
            header_image_url
            release_date
            primary_artist_id
            average_rating
            number_of_ratings
        }
    }
`

/**
 * GraphQL query: GET_SONGS_ON_TITLE.
 *
 * This query fetches a list of songs based on a search term, with optional parameters for pagination and sorting.
 * The details returned for each song include the header image URL, ID, title, artist names, average rating, number of ratings, and release date.
 *
 * @param {string} [$title] - The title or part of the title of the songs to search for.
 * @param {number} [$limit] - The maximum number of songs to return.
 * @param {string} [$sort] - The sorting criteria (e.g., 'rating', 'alphabetical').
 * @param {number} [$page] - The page number for pagination.
 * @returns {Array<Object>} An array of objects, each containing information about a song.
 */
export const GET_SONGS_ON_TITLE = gql`
    query ($title: String, $limit: Int, $sort: String, $page: Int) {
        getSongsOnTitle(
            title: $title
            limit: $limit
            sort: $sort
            page: $page
        ) {
            header_image_url
            id
            title
            artist_names
            average_rating
            number_of_ratings
            release_date
        }
    }
`

/**
 * GraphQL query: COUNT_SONGS.
 *
 * This query counts the number of songs in the database that match a given title or search term.
 *
 * @param {string} [$title] - The title or part of the title of the songs to count.
 * @returns {number} The count of songs matching the search criteria.
 */
export const COUNT_SONGS = gql`
    query ($title: String) {
        countSongs(title: $title)
    }
`
