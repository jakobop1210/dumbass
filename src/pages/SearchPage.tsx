import { useState } from "react";
import Navbar from "../components/molecules/Navbar";
import SearchBar from "../components/molecules/SearchBar";
import Filter from "../components/molecules/Sort";

function SearchPage(){
    const [filter, setFilter] = useState('Song');

    return (
        <div>
            <Navbar />
            <div className="flex justify-center">
                <SearchBar
                    className="w-4/5 mt-10 drop-shadow mb-10"
                    filterOptions={['Song', 'Artist']}
                    selectedFilter={filter}
                    onFilterChange={(newFilter) => setFilter(newFilter)}
                />
            </div>
            <div className="flex justify-center">
                <Filter />
            </div>
        </div>
    );
};

export default SearchPage;
