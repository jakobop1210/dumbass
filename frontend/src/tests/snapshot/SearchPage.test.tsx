import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { MockedProvider } from '@apollo/client/testing';
import SearchPage from '../../components/pages/SearchPage';
import { COUNT_ARTISTS, GET_ARTISTS_ON_NAME } from '../../graphql/queries/artistQueries';
import { COUNT_SONGS, GET_SONGS_ON_TITLE } from '../../graphql/queries/songQueries';

const mocks = [
    {
        request: {
            query: GET_ARTISTS_ON_NAME,
            variables: {
                name: 'Test Artist',
                limit: 12,
                sort: 'rating',
                page: 1
            },
        },
        result: {
            data: {
                getArtistsOnName: [
                    {
                        id: 1,
                        name: 'Test Artist',
                        alternate_names: ['Test Artist Alias'],
                        image_url: 'http://example.com/artist.jpg',
                        average_rating: 4.5,
                        number_of_ratings: 100
                    },
                ],
            },
        },
    },
    {
        request: {
            query: GET_SONGS_ON_TITLE,
            variables: {
                title: null,
                limit: 12,
                sort: 'relevance',
                page: 1
            },
        },
        result: {
            data: {
                getSongsOnTitle: [
                    {
                        header_image_url: 'http://example.com/song.jpg',
                        id: 1,
                        title: 'Test Song',
                        artist_names: ['Test Artist'],
                        average_rating: 4.2,
                        number_of_ratings: 80,
                        release_date: '2022-01-01'
                    },
                ],
            },
        },
    },
    {
        request: {
            query: COUNT_ARTISTS,
            variables: {
                name: 'Test Artist'
            },
        },
        result: {
            data: {
                countArtists: 0
            },
        },
    },
    {
        request: {
            query: COUNT_SONGS,
            variables: {
                title: null
            },
        },
        result: {
            data: {
                countSongs: 15
            },
        },
    },
];

test('SearchPage snapshot', () => {
    const { asFragment } = render(
        <MockedProvider mocks={mocks} addTypename={false}>
            <Router>
                <SearchPage />
            </Router>
        </MockedProvider>
    );
    expect(asFragment()).toMatchSnapshot();
});
