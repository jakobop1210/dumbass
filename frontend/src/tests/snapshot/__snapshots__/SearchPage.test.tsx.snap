// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchPage snapshot 1`] = `
<DocumentFragment>
  <main
    class="w-full"
  >
    <div
      class="w-1/2 text-left absolute"
    >
      <nav
        aria-label="breadcrumb"
        class="w-max"
      >
        <ol
          class="flex w-full flex-wrap items-center rounded-md bg-opacity-60 py-2 px-4 "
        >
          <li
            class="flex items-center font-sans text-sm font-normal leading-normal antialiased transition-colors"
          >
            <a
              class="opacity-60 duration-300 hover:text-green"
              href="/"
            >
              <svg
                class="h-4 w-4"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10.707 2.293a1 1 0 00-1.414 0l-7 7a1 1 0 001.414 1.414L4 10.414V17a1 1 0 001 1h2a1 1 0 001-1v-2a1 1 0 011-1h2a1 1 0 011 1v2a1 1 0 001 1h2a1 1 0 001-1v-6.586l.293.293a1 1 0 001.414-1.414l-7-7z"
                />
              </svg>
            </a>
            <span
              class="pointer-events-none mx-2 select-none font-sans text-sm font-normal leading-normal antialiased"
            >
              /
            </span>
          </li>
          <li
            class="flex items-center font-sans text-sm font-normal leading-normal antialiased transition-colors"
          >
            <div
              class="opacity-70"
            >
              <span>
                Search
              </span>
            </div>
          </li>
        </ol>
      </nav>
    </div>
    <header
      class="flex justify-center"
    >
      <div
        class="relative w-11/12 sm:w-3/5 mt-10 drop-shadow mb-5 z-40"
      >
        <form
          autocomplete="off"
          class="flex items-center pl-2 pr-2 bg-[#FFFFFF] rounded-lg h-14"
          role="search"
        >
          <label
            class="sr-only"
            for="searchInput"
          >
            Search for a song or artist
          </label>
          <input
            class="w-full p-2 outline-none rounded-md"
            id="searchInput"
            placeholder="Search for a song or artist"
            type="text"
            value=""
          />
          <div
            class="h-full border-l-2 flex justify-center items-center"
          >
            <div
              class="relative inline-block text-left h-full"
              data-headlessui-state=""
            >
              <div
                class="absolute -top-5 text-xs mb-1 opacity-70"
              >
                Filter by
              </div>
              <section
                class="h-full w-full"
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="menu"
                  class="w-24 inline-flex justify-center items-center h-full px-4 py-2 bg-white text-sm font-medium text-gray-700 hover:bg-gray-50  focus-visible:ring-2 focus-visible:ring-white focus-visible:ring-opacity-75"
                  data-headlessui-state=""
                  id="headlessui-menu-button-:r0:"
                  type="button"
                >
                  <span
                    class="block overflow-hidden text-ellipsis whitespace-nowrap"
                  >
                    Song
                  </span>
                  <svg
                    aria-hidden="true"
                    class="ml-2 -mr-1 h-5 w-5"
                    fill="currentColor"
                    viewBox="0 0 20 20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M5.23 7.21a.75.75 0 011.06.02L10 11.168l3.71-3.938a.75.75 0 111.08 1.04l-4.25 4.5a.75.75 0 01-1.08 0l-4.25-4.5a.75.75 0 01.02-1.06z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </button>
              </section>
            </div>
          </div>
          <button
            aria-label="Search"
            class="p-2 rounded-md ml-2"
            data-testid="search-button"
            type="submit"
          >
            <svg
              color="#999"
              data-testid="search-icon"
              fill="currentColor"
              height="20"
              stroke="currentColor"
              stroke-width="0"
              style="color: rgb(153, 153, 153);"
              viewBox="0 0 512 512"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
              />
            </svg>
          </button>
        </form>
      </div>
    </header>
    <section
      class="flex text-xl justify-center w-full text-gray-500"
    >
      Please enter a search term...
    </section>
  </main>
</DocumentFragment>
`;
